### ISSUE TITLE
Idea: Replace history planes with auxiliary outputs/targets for previous moves/states 

### ISSUE THREAD

**Issue by Hersmunch** [lightvector/KataGo#94]
So a had an idea that perhaps adding auxiliary outputs/targets for previous the previous n moves and/or states. I think doing so might make the use of historical board states as inputs redundant as the weights would have to work them out instead, including picking up on ko from the lack of double retakes in training data (though supreko might be too rare to understand?).

Anyway, I hadn't seen this mentioned anywhere before - any idea if it has been thought about or experimented with? Perhaps it could be added to the list of (presumably many) things to study with katago? :)

I think it would be interesting to see how strength varies with how many moves back are used during training and comparing that to having history and not having any history or auxiliary target(s). I wonder how potentially having some knowledge about the "flow of the game" might change chosen moves too.

I can't immediately see how the extra target(s) could be used in the MCTS utility function but a more out there idea might be to see if the net itself could make use of it by having a trained weighting? Maybe.. that can wait :)

Any other thoughts welcome. Thank you in advance!

**Comment by lightvector** [comment]
I'm skeptical that this will be useful. When looking for additional targets to predict, generally you want those targets to have a meaningful correlation with the things you actually care about, so that they can serve as data-rich regularization for those things. Ownership and score are obviously highly correlated with the game winner. Future opponent or self moves are obviously highly correlated with urgent areas to play, and very directly correlated with moves that you might want to play right now (for example, [1-2-3 heuristic](https://senseis.xmp.net/?OneTwoThree)). Predicting a move that you want to play next turn gives you very strong information about where you want to play now.

The set of plausible ways to reach a given position (including plausible ways that are not exactly the move sequence that was actually played in a game) is seems less related to things you care about. So my inclination is to think that it is not as likely to be useful.

Keep in mind that predicting a target is not in general the same a providing it as an input. They have very different effects, and in general to predict something does NOT make it redundant to input. For example, (stupid extreme example) - it would not be possible to make the "black stone input plane" redundant by only telling the neural net where the white stones were and having it predict where the black stones were. 

If you would like to experiment with this yourself, please feel free. But don't jump too far ahead - I don't know of anybody that has even done a rigorous controlled test of history versus no history. Perhaps best to start with such a test first, before adding more targets!

**Comment by Hersmunch** [comment]
Thanks for taking the time to consider the idea (I'm sure it can be exhausting explaining things to "novices" like myself). You make some constructive and convincing arguments too :)
I think I had trouble in the past attempting to install tensorflow as I only have an AMD GPU, which I believe means I would need to train remotely?

In terms of potential ways in which predicting these auxiliary targets could be useful as part of learning might be helping with policy. For example, learning the exact order of moves to take apart a puzzle and reversing it allows the puzzle to be completed, or in our case, sequences of moves predicted.

If the board history input planes are kept and the extra targets added, would that be a stupid thing to do? I mean, it would be trivial to map the inputs straight to the outputs. Would that still be the case if there are fewer actual history inputs states than predicted previous states/moves?

I did see your previous experiments with having no history here https://github.com/lightvector/GoNN#some-thoughts-about-history-as-an-input-mar-2018 - is there any newer/other that you know of?
Also, any tips on how one might think about constructing an appropriate loss function for previous moves/states would be appreciated.


Another related but probably more out there idea would be to instead provide some extra "feedback" from the policy and/or board state outputs as inputs to the next state's evaluation (alongside the actual board state history). I'm thinking this would allow for the difference between them to be considered during evaluation with the aim of providing feedback on how predictable the opponents moves are. This might allow for dynamic change of "tactics"/policy consideration?
There are a lot of ifs in there so it does seem a bit far-fetched. Also, might this already be happening e.g. a set of weights might be using the all but the most recent of the history states to predict what the most recent state is and then compare that to the actual input? How might one try to test if that is happening?

So er,  thanks again :)

**Comment by lightvector** [comment]
Digging through old issues, apparently never replied back here. :)

* Yes, if you can't install Tensorflow it will be hard to train. Renting cloud machines is a possibility, I do a lot of dev nowadays on the cloud since my laptop has no real GPU at all!
* Yes, I think there is usually no real point to having outputs that are the same as inputs, but relatedly you do do that sort of thing when trying to train autoencoders. (https://en.wikipedia.org/wiki/Autoencoder)
* I haven't experimented further with having no history, but actually SAI's (another distributed project) whole training run has no history in it. Experimentally, disabling history in KataGo drops the Elo by something like 400 points, but possibly that's because KataGo is trained with history. Or rather, all but a few percent of samples are trained with history - KataGo does mask history on a few samples so as to guarantee robustness on historyless positions (e.g. whole-board tsumego and such).
* If you like to try learning more, go rent a cloud machine or set yourself up on Google colab, learn Tensorflow and start running some experiments like I did in GoNN. :)
If you turn off a cloud GPU machine when not using it, it's not that expensive. 60 cents per hour can get you something basic and working for small net training already. And Google colab should be free for the first 300$, or something like that.
